### PR TITLE
SOAPUIOS-360 Implement save request action

### PR DESCRIPTION
**Pull request by omitina**
https://smartbear.atlassian.net/browse/SOAPUIOS-360

### CODE REVIEW

File: soapui/src/main/java/com/eviware/soapui/impl/actions/RestServiceBuilder.java [hunk 1]
```diff
@@ -64,27 +67,79 @@ public String getUri() {
         }
     }
 
-    public void createRestService(WsdlProject project, String URI) throws MalformedURLException {
+    public RestRequest createRestService(WsdlProject project, String URI) throws MalformedURLException {
```
Comment (IlyaAvdeev) [inline soapui/src/main/java/com/eviware/soapui/impl/actions/RestServiceBuilder.java]: why createRestService returns RestRequest?

Comment (IlyaAvdeev) [reply soapui/src/main/java/com/eviware/soapui/impl/actions/RestServiceBuilder.java]: I guess this public class being used by 3rd party plugins, so changing public interface breaks backward compatibility.

File: soapui/src/main/java/com/eviware/soapui/impl/rest/actions/method/SaveRequestAction.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+import javax.swing.AbstractAction;
+import javax.swing.JPanel;
+import java.awt.BorderLayout;
+import java.awt.Dimension;
+import java.awt.event.ActionEvent;
+import java.util.HashMap;
+import java.util.List;
+
+public class SaveRequestAction extends AbstractAction {
+    public static final MessageSupport messages = MessageSupport.getMessages(SaveRequestAction.class);
```
Comment (IlyaAvdeev) [inline soapui/src/main/java/com/eviware/soapui/impl/rest/actions/method/SaveRequestAction.java]: why public?

File: soapui/src/main/java/com/eviware/soapui/impl/rest/actions/method/SaveRequestAction.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    private static final String NEW_PROJECT_OPTION = "[new project]";
+    private static final String DEFAULT_PROJECT_NAME = "Project";
+    private static final int DIALOG_WIDTH = 520;
+    private static final int DIALOG_HEIGHT = 300;
+    private static final Dimension PROJECTS_FORM_SIZE = new Dimension(400, 100);
+    private XFormDialog dialog;
+    private ProjectListItem selectedProject;
+    private HashMap<String, Object> context;
```
Comment (IlyaAvdeev) [inline soapui/src/main/java/com/eviware/soapui/impl/rest/actions/method/SaveRequestAction.java]: I guess Map in the declaration is enough, HashMap is detail of implementation

File: soapui/src/main/java/com/eviware/soapui/impl/rest/actions/method/SaveRequestAction.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+    public SaveRequestAction(HashMap<String, Object> context) {
```
Comment (IlyaAvdeev) [inline soapui/src/main/java/com/eviware/soapui/impl/rest/actions/method/SaveRequestAction.java]: Map

File: soapui/src/main/java/com/eviware/soapui/impl/actions/RestServiceBuilder.java [hunk 5]
```diff
@@ -64,27 +67,79 @@ public String getUri() {
...
+    public RestRequest createRestServiceHeadlessFromUri(WsdlProject project, RequestInfo requestInfo,
+                                                        ModelCreationStrategy methodReuseStrategy) throws MalformedURLException {
         RestResource restResource = createResource(REUSE_MODEL, project, requestInfo.getUri());
         RestMethod restMethod = addNewMethod(methodReuseStrategy, restResource, requestInfo.getRequestMethod());
         RestRequest restRequest = addNewRequest(restMethod);
         copyParametersWithDefaultsOnResource(extractParams(requestInfo.getUri()), restMethod.getParams(), restRequest.getParams());
         return restRequest;
     }
 
+    public RestRequest createRestServiceWithMethod(WsdlProject project, String URI,
```
Comment (IlyaAvdeev) [inline soapui/src/main/java/com/eviware/soapui/impl/actions/RestServiceBuilder.java:92]: should this method be public, according to the signature it looks like private helper method.

File: soapui/src/main/java/com/eviware/soapui/impl/actions/RestServiceBuilder.java [hunk 6]
```diff
@@ -64,27 +67,79 @@ public String getUri() {
...
+        }
+        copyParameters(extractParams(URI), restMethod.getParams());
+        if (showDesktopPanel) {
+            UISupport.select(restRequest);
+            UISupport.showDesktopPanel(restRequest);
+        }
+        return restRequest;
+    }
+
+    public RestRequest createRestServiceFromInspectionData(WsdlProject project, String URI,
```
Comment (IlyaAvdeev) [inline soapui/src/main/java/com/eviware/soapui/impl/actions/RestServiceBuilder.java:121]: do you call this method externally? should it be public?

**Review by IlyaAvdeev — CHANGES_REQUESTED** [review]

File: soapui/src/main/java/com/eviware/soapui/impl/actions/RestServiceBuilder.java (hunk 5 above)
Comment (omitina) [reply soapui/src/main/java/com/eviware/soapui/impl/actions/RestServiceBuilder.java:92]: I use it in SaveRequestAction class

File: soapui/src/main/java/com/eviware/soapui/impl/actions/RestServiceBuilder.java (hunk 6 above)
Comment (omitina) [reply soapui/src/main/java/com/eviware/soapui/impl/actions/RestServiceBuilder.java:121]: yes, in SaveRequestAction class

File: soapui/src/main/java/com/eviware/soapui/impl/actions/RestServiceBuilder.java (hunk 1 above)
Comment (omitina) [reply soapui/src/main/java/com/eviware/soapui/impl/actions/RestServiceBuilder.java]: Agree

File: soapui/src/main/java/com/eviware/soapui/impl/rest/actions/method/SaveRequestAction.java (hunk 2 above)
Comment (omitina) [reply soapui/src/main/java/com/eviware/soapui/impl/rest/actions/method/SaveRequestAction.java]: changed

File: soapui/src/main/java/com/eviware/soapui/impl/rest/actions/method/SaveRequestAction.java (hunk 3 above)
Comment (omitina) [reply soapui/src/main/java/com/eviware/soapui/impl/rest/actions/method/SaveRequestAction.java]: Agree

File: soapui/src/main/java/com/eviware/soapui/impl/rest/actions/method/SaveRequestAction.java (hunk 4 above)
Comment (omitina) [reply soapui/src/main/java/com/eviware/soapui/impl/rest/actions/method/SaveRequestAction.java]: changed

**Review by IlyaAvdeev — APPROVED** [review]

**Comment by IlyaAvdeev** [comment]
Looks good, well done!

**Pull request merged**